src/transformers/trainer_pt_utils.py

```diff
@@ -311,40 +320,64 @@ class EvalLoopContainer:
-        """Add tensors to the stored objects. If `do_nested_concat=True`, the tensors will be concatenated recursively."""
-        if self.tensors is None:
-            self.tensors = tensors if self.do_nested_concat else [tensors]
+    def add(self, device_tensors) -> None:
+        """Add device_tensors to the stored objects. If `do_nested_concat=True`, the tensors will be concatenated
+        recursively, otherwise they will be stored in a list."""
+        if device_tensors is None:
+            return
+
+        # If `do_nested_concat=True` self.device_tensors will be a nested structure of tensors
         elif self.do_nested_concat:
-            self.tensors = nested_concat(self.tensors, tensors, padding_index=self.padding_index)
+            if self.device_tensors is None:
+                self.device_tensors = device_tensors
+            else:
+                self.device_tensors = nested_concat(
+                    self.device_tensors, device_tensors, padding_index=self.padding_index
+                )
+
+        # If `do_nested_concat=False` self.device_tensors will be a list of tensors
         else:
-            self.tensors.append(tensors)
+            if self.device_tensors is None:
+                self.device_tensors = [device_tensors]
+            else:
+                self.device_tensors.append(device_tensors)
 
-    def to_cpu_and_numpy(self) -> None:
-        """Move tensors in stored objects to CPU and convert them to numpy arrays."""
+    def cpu(self) -> None:
+        """Move all stored tensors to CPU (inplace)."""
 
         # Check if we have something to add, if not just return
-        if self.tensors is None:
+        if self.device_tensors is None:
             return
 
-        new_arrays = nested_numpify(self.tensors)
-        if self.arrays is None:
-            self.arrays = new_arrays
+        new_cpu_data = nested_cpu(self.device_tensors)
+        if self.cpu_tensors is None:
+            self.cpu_tensors = new_cpu_data
         elif self.do_nested_concat:
-            self.arrays = nested_concat(self.arrays, new_arrays, padding_index=self.padding_index)
+            self.cpu_tensors = nested_concat(self.cpu_tensors, new_cpu_data, padding_index=self.padding_index)
         else:
-            self.arrays.extend(new_arrays)
+            self.cpu_tensors.extend(new_cpu_data)
 
         # reset device tensors after adding to cpu
-        self.tensors = None
+        self.device_tensors = None
+
+    def get(self, to_numpy: bool = False) -> Any:
+        """Return all the stored tensors objects. All tensors are located to CPU device.
+
+        Args:
+            to_numpy (`bool`, *optional*, defaults to `False`):
+                Whether to convert the tensors to numpy format.
+
+        Returns:
+            The stored objects with torch tensors or numpy arrays. If no tensors are stored, returns `None`.
+        """
+        if self.device_tensors is not None:
+            self.cpu()
```

Comment:
add a comment to say that we make sure that all tensors have been moved to the cpu.

Reply:
Added in https://github.com/huggingface/transformers/pull/31388/commits/7d18eb67a5541719b9ea829012927f288df645b8